<a href="https://colab.research.google.com/github/Xavier-ML/BIOINFORMATICA-XAVIER-ML/blob/main/Examen_Final_Preguntas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Examen Final - Python Fundamentals for Data Science
Instrucciones:
* El examen tiene un máximo de 2.5 horas de duración
* Se permite hacer consultas por internet
* Las respuestas incluyendo código deben verse reflejadas en el notebook
* Enviar el notebook con respuestas al correo del instructor al terminar
* No se permite consultar ni conversar con los compañeros de clase

P1. Cargar el dataset datos.csv en Pandas.
Mostrar las filas iniciales en la pantalla. Cuantas filas y columnas hay

In [37]:
# Cargar el dataset
import pandas as pd
datos = pd.read_csv("/content/datos.csv", sep = ",")

In [38]:
# Mostrar filas iniciales en la pantalla
datos.head(10)

Unnamed: 0  age         workclass  fnlwgt     education  education-num  \
0           0   50  Self-emp-not-inc   83311     Bachelors             13   
1           1   38           Private  215646       HS-grad              9   
2           2   53           Private  234721          11th              7   
3           3   28           Private  338409     Bachelors             13   
4           4   37           Private  284582       Masters             14   
5           5   49           Private  160187           9th              5   
6           6   52  Self-emp-not-inc  209642       HS-grad              9   
7           7   31           Private   45781       Masters             14   
8           8   42           Private  159449     Bachelors             13   
9           9   37           Private  280464  Some-college             10   

          marital-status         occupation   relationship   race     sex  \
0     Married-civ-spouse    Exec-managerial        Husband  White    Male   
1               Divorced  Handlers-cleaners  Not-in-family  White    Male   
2     Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
3     Married-civ-spouse     Prof-specialty           Wife  Black  Female   
4     Married-civ-spouse    Exec-managerial           Wife  White  Female   
5  Married-spouse-absent      Other-service  Not-in-family  Black  Female   
6     Married-civ-spouse    Exec-managerial        Husband  White    Male   
7          Never-married     Prof-specialty  Not-in-family  White  Female   
8     Married-civ-spouse    Exec-managerial        Husband  White    Male   
9     Married-civ-spouse    Exec-managerial        Husband  Black    Male   

   capital-gain  capital-loss  hours-per-week native-country salary  
0             0             0              13  United-States  <=50K  
1             0             0              40  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40           Cuba  <=50K  
4             0             0              40  United-States  <=50K  
5             0             0              16        Jamaica  <=50K  
6             0             0              45  United-States   >50K  
7         14084             0              50  United-States   >50K  
8          5178             0              40  United-States   >50K  
9             0             0              80  United-States   >50K

In [39]:
#Cuantas filas y columnas hay
datos.shape

(32560, 16)

P2. Existe una columna que copia el índice. Eliminarla del dataframe y mostrar el resultado.

In [40]:
datos.columns

Index(['Unnamed: 0', 'age', 'workclass', 'fnlwgt', 'education',
       'education-num', 'marital-status', 'occupation', 'relationship', 'race',
       'sex', 'capital-gain', 'capital-loss', 'hours-per-week',
       'native-country', 'salary'],
      dtype='object')

In [41]:
datos = datos.drop('Unnamed: 0', axis=1)
datos.head(10)

age         workclass  fnlwgt     education  education-num  \
0   50  Self-emp-not-inc   83311     Bachelors             13   
1   38           Private  215646       HS-grad              9   
2   53           Private  234721          11th              7   
3   28           Private  338409     Bachelors             13   
4   37           Private  284582       Masters             14   
5   49           Private  160187           9th              5   
6   52  Self-emp-not-inc  209642       HS-grad              9   
7   31           Private   45781       Masters             14   
8   42           Private  159449     Bachelors             13   
9   37           Private  280464  Some-college             10   

          marital-status         occupation   relationship   race     sex  \
0     Married-civ-spouse    Exec-managerial        Husband  White    Male   
1               Divorced  Handlers-cleaners  Not-in-family  White    Male   
2     Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
3     Married-civ-spouse     Prof-specialty           Wife  Black  Female   
4     Married-civ-spouse    Exec-managerial           Wife  White  Female   
5  Married-spouse-absent      Other-service  Not-in-family  Black  Female   
6     Married-civ-spouse    Exec-managerial        Husband  White    Male   
7          Never-married     Prof-specialty  Not-in-family  White  Female   
8     Married-civ-spouse    Exec-managerial        Husband  White    Male   
9     Married-civ-spouse    Exec-managerial        Husband  Black    Male   

   capital-gain  capital-loss  hours-per-week native-country salary  
0             0             0              13  United-States  <=50K  
1             0             0              40  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40           Cuba  <=50K  
4             0             0              40  United-States  <=50K  
5             0             0              16        Jamaica  <=50K  
6             0             0              45  United-States   >50K  
7         14084             0              50  United-States   >50K  
8          5178             0              40  United-States   >50K  
9             0             0              80  United-States   >50K

P3. ¿Cuántas personas de sexo masculino y femenino están representadas en el dataset?

In [42]:
datos.groupby(["sex"])["sex"].count()

sex
Female    10771
Male      21789
Name: sex, dtype: int64

In [43]:
10771 + 21789

32560

P4. ¿Cuál es la edad promedio del sexo masculino?

In [44]:
datos.groupby(["sex"])["age"].mean()

sex
Female    36.858230
Male      39.433567
Name: age, dtype: float64

promedio del sexo masculino es 39.43 redondeando a dos cifras 39 la edad.

P5. ¿Cuál es la proporción de Mexicanos del dataset?



In [47]:
datos_unicos_country = datos['native-country'].unique()
datos_unicos_country

array(['United-States', 'Cuba', 'Jamaica', 'India', '?', 'Mexico',
       'South', 'Puerto-Rico', 'Honduras', 'England', 'Canada', 'Germany',
       'Iran', 'Philippines', 'Italy', 'Poland', 'Columbia', 'Cambodia',
       'Thailand', 'Ecuador', 'Laos', 'Taiwan', 'Haiti', 'Portugal',
       'Dominican-Republic', 'El-Salvador', 'France', 'Guatemala',
       'China', 'Japan', 'Yugoslavia', 'Peru',
       'Outlying-US(Guam-USVI-etc)', 'Scotland', 'Trinadad&Tobago',
       'Greece', 'Nicaragua', 'Vietnam', 'Hong', 'Ireland', 'Hungary',
       'Holand-Netherlands'], dtype=object)

In [48]:
Mexicanos = datos[datos["native-country"] == "Mexico"]
Mexicanos

age  workclass  fnlwgt     education  education-num  \
14      34    Private  245487       7th-8th              4   
55      46    Private  216666       5th-6th              3   
74      27    Private  213921       HS-grad              9   
141     19    Private  316868  Some-college             10   
149     47  Local-gov  186009  Some-college             10   
...    ...        ...     ...           ...            ...   
32331   26    Private  233777       HS-grad              9   
32333   24    Private  176580       5th-6th              3   
32424   23    Private  180771       1st-4th              2   
32431   36    Private  208068     Preschool              1   
32546   43    Private  260761       HS-grad              9   

              marital-status         occupation   relationship  \
14        Married-civ-spouse   Transport-moving        Husband   
55        Married-civ-spouse  Machine-op-inspct        Husband   
74             Never-married      Other-service      Own-child   
141            Never-married      Other-service      Own-child   
149                 Divorced       Adm-clerical      Unmarried   
...                      ...                ...            ...   
32331     Married-civ-spouse       Craft-repair        Husband   
32333  Married-spouse-absent    Farming-fishing  Not-in-family   
32424     Married-civ-spouse  Machine-op-inspct           Wife   
32431               Divorced      Other-service  Not-in-family   
32546     Married-civ-spouse  Machine-op-inspct        Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
14     Amer-Indian-Eskimo    Male             0             0              45   
55                  White    Male             0             0              40   
74                  White    Male             0             0              40   
141                 White    Male             0             0              30   
149                 White  Female             0             0              38   
...                   ...     ...           ...           ...             ...   
32331               White    Male             0             0              72   
32333               White    Male             0             0              40   
32424  Amer-Indian-Eskimo  Female             0             0              35   
32431               Other    Male             0             0              72   
32546               White    Male             0             0              40   

      native-country salary  
14            Mexico  <=50K  
55            Mexico  <=50K  
74            Mexico  <=50K  
141           Mexico  <=50K  
149           Mexico  <=50K  
...              ...    ...  
32331         Mexico  <=50K  
32333         Mexico  <=50K  
32424         Mexico  <=50K  
32431         Mexico  <=50K  
32546         Mexico  <=50K  

[643 rows x 15 columns]

In [51]:
Mexicanos.shape

(643, 15)

In [50]:
(643/32560)*100

1.9748157248157248

El 1.97% del dataset son Mexicanos.

P6. ¿Cuál es la edad promedio de aquellos que ganan más de 50K y menos de 50K?

In [58]:
gananmayora50k = datos[datos.salary == ">50K"]
gananmayora50k

age         workclass  fnlwgt     education  education-num  \
6       52  Self-emp-not-inc  209642       HS-grad              9   
7       31           Private   45781       Masters             14   
8       42           Private  159449     Bachelors             13   
9       37           Private  280464  Some-college             10   
10      30         State-gov  141297     Bachelors             13   
...    ...               ...     ...           ...            ...   
32538   71                 ?  287372     Doctorate             16   
32544   39         Local-gov  111499    Assoc-acdm             12   
32553   53           Private  321865       Masters             14   
32556   40           Private  154374       HS-grad              9   
32559   52      Self-emp-inc  287927       HS-grad              9   

           marital-status         occupation   relationship  \
6      Married-civ-spouse    Exec-managerial        Husband   
7           Never-married     Prof-specialty  Not-in-family   
8      Married-civ-spouse    Exec-managerial        Husband   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse     Prof-specialty        Husband   
...                   ...                ...            ...   
32538  Married-civ-spouse                  ?        Husband   
32544  Married-civ-spouse       Adm-clerical           Wife   
32553  Married-civ-spouse    Exec-managerial        Husband   
32556  Married-civ-spouse  Machine-op-inspct        Husband   
32559  Married-civ-spouse    Exec-managerial           Wife   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
6                   White    Male             0             0              45   
7                   White  Female         14084             0              50   
8                   White    Male          5178             0              40   
9                   Black    Male             0             0              80   
10     Asian-Pac-Islander    Male             0             0              40   
...                   ...     ...           ...           ...             ...   
32538               White    Male             0             0              10   
32544               White  Female             0             0              20   
32553               White    Male             0             0              40   
32556               White    Male             0             0              40   
32559               White  Female         15024             0              40   

      native-country salary  
6      United-States   >50K  
7      United-States   >50K  
8      United-States   >50K  
9      United-States   >50K  
10             India   >50K  
...              ...    ...  
32538  United-States   >50K  
32544  United-States   >50K  
32553  United-States   >50K  
32556  United-States   >50K  
32559  United-States   >50K  

[7841 rows x 15 columns]

In [61]:
gananmenora50k = datos[datos.salary == "<50K"]
gananmenora50k

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, salary]
Index: []

In [60]:
gananmayora50k.groupby(["salary"])["age"].mean()

salary
>50K    44.249841
Name: age, dtype: float64

la edad promedio es 44 años.

P7. ¿Todos los que ganan mas de 50K al año tienen un Bachiller?

In [62]:
gananmayora50k.shape

(7841, 15)

In [64]:
Sonbachiller = gananmayora50k[gananmayora50k["education"] == "Bachelors"]
Sonbachiller

age     workclass  fnlwgt  education  education-num  \
8       42       Private  159449  Bachelors             13   
10      30     State-gov  141297  Bachelors             13   
24      56     Local-gov  216851  Bachelors             13   
44      57   Federal-gov  337895  Bachelors             13   
52      50   Federal-gov  251585  Bachelors             13   
...    ...           ...     ...        ...            ...   
32463   45  Self-emp-inc  173664  Bachelors             13   
32529   35             ?  320084  Bachelors             13   
32532   54       Private  337992  Bachelors             13   
32535   34       Private  160216  Bachelors             13   
32537   38       Private  139180  Bachelors             13   

           marital-status       occupation   relationship                race  \
8      Married-civ-spouse  Exec-managerial        Husband               White   
10     Married-civ-spouse   Prof-specialty        Husband  Asian-Pac-Islander   
24     Married-civ-spouse     Tech-support        Husband               White   
44     Married-civ-spouse   Prof-specialty        Husband               Black   
52               Divorced  Exec-managerial  Not-in-family               White   
...                   ...              ...            ...                 ...   
32463  Married-civ-spouse  Exec-managerial        Husband               White   
32529  Married-civ-spouse                ?           Wife               White   
32532  Married-civ-spouse  Exec-managerial        Husband  Asian-Pac-Islander   
32535       Never-married  Exec-managerial  Not-in-family               White   
32537            Divorced   Prof-specialty      Unmarried               Black   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
8        Male          5178             0              40  United-States   
10       Male             0             0              40          India   
24       Male             0             0              40  United-States   
44       Male             0             0              40  United-States   
52       Male             0             0              55  United-States   
...       ...           ...           ...             ...            ...   
32463    Male             0             0              45  United-States   
32529  Female             0             0              55  United-States   
32532    Male             0             0              50          Japan   
32535  Female             0             0              55  United-States   
32537  Female         15020             0              45  United-States   

      salary  
8       >50K  
10      >50K  
24      >50K  
44      >50K  
52      >50K  
...      ...  
32463   >50K  
32529   >50K  
32532   >50K  
32535   >50K  
32537   >50K  

[2221 rows x 15 columns]

In [65]:
7841 - 2221

5620

No, Solo 2221 tienen bachiller y los 5620 otros grados que ganan más de 50k.

P8. Determinar el número de horas promedio por semana que se trabaja en cada país para <=50K y >50K

P9. Agrupar y contar los registros de <=50K y >50K. Graficar el resultado de manera apropiada

P10. Según el análisis:
* ¿Qué campos serían idóneos para entrenar un modelo que predice si nuevos datos corresponden a <=50K o >50K?
* ¿Sería un modelo de regresión o clasificación - por qué?
* Argumentar en máx 3 párrafos.

P11.En un examen tipo test, las preguntas correctas suman un punto y las incorrectas restan medio punto. En total hay 100 preguntas y no se admiten respuestas en blanco (hay que contestar todas).

La nota de un alumno es 8.05 sobre 10. Calcular el número de preguntas que contestó correcta e incorrectamente.

P12. Si se suma 7 al numerador y al denominador de una determinada fracción, se obtiene la fracción

2
3
Si en vez de sumar 7 se resta 3 al numerador y al denominador, se obtiene la fracción

1
4
Encontrar dicha fracción.

P13. Graficar Histograma de rangos de edades

P14.Graficar Violin  Hours per week

P15.Graficar dispersión entre edad y horas per week

P16.Una marca de bebidas prepara una limonada (agua y concentrado de limón) con una cantidad muy precisa de sus ingredientes. La relación entre las cantidades de agua y concentrado limón es

L
L
=
2
L
A
5
donde LL representa los litros de concentrado de limón y LA los litros de agua.

Si se necesitan 20 limones para obtener un litro de concentrado de limón, ¿cuántos limones se necesitan para elaborar 1230 botellas de 2L de esta limonada?

P17.Grafico de barras de Race

P18. Desarrollar un programa que te pregunte un número y termine calculando el factorial, ejemplo= 3  ---> 1*2*3 =6

P19.Desarrollar un programa que te calcule la nota final del curso, aoartir de las notas de  tareas, laboratorios y examen final (0.3 ,0.4, 0.3 pesos respectivamente)

P20. 3 resumenes de la data , seleccionar mínimo 2 variables cuantitativas